build/main.rs

```diff
@@ -166,28 +202,34 @@ fn main() {
 		utils::optimize(
 			&mut module,
 			vec![CALL_SYMBOL]
-		).expect("Optimizer to finish without errors");
+		).unwrap_or_else(|e| die(Error::from(e)))
 	}
 
 	if let Some(save_raw_path) = matches.value_of("save_raw") {
 		parity_wasm::serialize_to_file(save_raw_path, module.clone())
-			.expect("Failed to write intermediate module");
+			.unwrap_or_else(|e| die(Error::Encoding(e)));
 	}
 
-	let raw_module = parity_wasm::serialize(module).expect("Failed to serialize module");
+	let raw_module = parity_wasm::serialize(module)
+		.unwrap_or_else(|e| die(Error::Encoding(e)));
 
 	// If module has an exported function with name=CREATE_SYMBOL
 	// build will pack the module (raw_module) into this funciton and export as CALL_SYMBOL.
 	// Otherwise it will just save an optimised raw_module
 	if has_ctor(&ctor_module) {
 		if !matches.is_present("skip_optimization") {
-			utils::optimize(&mut ctor_module, vec![CREATE_SYMBOL]).expect("Optimizer to finish without errors");
+			utils::optimize(&mut ctor_module, vec![CREATE_SYMBOL])
+				.unwrap_or_else(|e| die(Error::from(e)))
 		}
-		let ctor_module = utils::pack_instance(raw_module, ctor_module).expect("Packing failed");
-		parity_wasm::serialize_to_file(&path, ctor_module).expect("Failed to serialize to file");
+		let ctor_module = utils::pack_instance(raw_module, ctor_module)
+			.unwrap_or_else(|e| die(Error::from(e)));
+		parity_wasm::serialize_to_file(&path, ctor_module)
+			.unwrap_or_else(|e| die(Error::Encoding(e)))
 	} else {
-		let mut file = fs::File::create(&path).expect("Failed to create file");
-		file.write_all(&raw_module).expect("Failed to write module to file");
+		let mut file = fs::File::create(&path)
+			.unwrap_or_else(|io| die(Error::from(io)));
+		file.write_all(&raw_module)
+			.unwrap_or_else(|io| die(Error::from(io)));
```

Comment:
All these `.unwrap_or_else(|_| die(...))` seems to me a bit confusing and boilerplaty.

I think it would be better to extract the body of this function into a new function, say `do_main() → Result<(), Error>`; then in `main` we just call it `die` if the result is `Err`; convert all these `unwrap_or_else` to `?`. What do you think?

Reply:
sure!